### ISSUE TITLE
Request-splitter module

### ISSUE THREAD

**Issue by nfeske** [cbe-project/cbe#4]
The request-splitter module canonicalizes client-originating requests into a sequence of block-sized primitives. To perform its operation, it does not need to touch any data. It only operates on the data offsets of the requests.

It implements the following interface:

**`request_acceptable`** returns true if there are no outstanding generated primitives. New work can be accepted.

**`submit_request`** loads the given request parameters into its internal state. (imagine importing a todo-list of N items where N corresponds to the request size).

**`peek_generated_primitive`** returns the next pending primitive (the current todo-list item). If there is no primitive generated, return a special primitive (`EMPTY`). _Alternatively, we may consider a separate function `generated_primitive_pending` that just return true/false._

**`take_generated_primitive`** advances the progress (offset, block number). It crosses the current item from the todo list.

E.g., if presented with a request "read 16 KiB starting from block 23", consecutive calls of the splitter's `generate_primitive` functions would produce four primitives "read block at 23", "read block at 24", "read block at 25", and "read block at 26".

**Comment by senier** [comment]
I'd lean towards the `generated_primitive_pending` approach, as it makes nicer preconditions and is symmetric with `request_acceptable`.

**Comment by nfeske** [comment]
To explain the reasoning behind the `peek_generated_primitive`, I expect that there are situations where the response of asking the next module in the pipeline for `primitive_acceptable` depends on the actual primitive not just the fact that there is a primitive.

For example, the next module may be fine with accepting a write request but cannot deal with a sync request at the time. So here, the boolean response of a `generated_primitive_pending` call is not sufficient to decide whether it is safe to take out a generated primitive from the splitter. Here we would first get the information about the next request in line from the splitter (peek) without changing the splitter's state. Then we would ask the next module if it can swallow such a request. Only when the next module is fine with that, we would "take" the request from the splitter (changing the splitter's state). Otherwise both the splitter's and the next module's state remain unchanged.

**Comment by senier** [comment]
I see. In this case the `peek` makes sense.

**Comment by m-stein** [comment]
I have a working prototype of the Splitter written in Ada:
https://github.com/m-stein/genode/commits/cbe
 
It took some time to get warm with it but I guess that the next steps come with much less friction. I wonder whether I should proceed with translating further modules or do formal verification of the Splitter module first?

**Comment by senier** [comment]
Great! I'd be happy to walk through the code in our next meeting.

Regarding next steps: You won't be verifying SPARK code all day long. IMHO it's sensible to implement and verify in parallel.

**Comment by m-stein** [comment]
The branch now proves at least all the inner splitter code (not all the glue code towards C++):
https://github.com/m-stein/genode/commits/cbe

**Comment by nfeske** [comment]
Thanks for posting the branch, @m-stein. It is very interesting to have a glimpse.

Regarding the size-consistency checking, let me propose of decoupling the checks from the construction of objects. This would cleanly separate both concerns and hopefully be a useful pattern to reuse.

* For each object type, there could be a SPARK function `Size_Of` that takes an object type as `in` argument and returns the `Object'Size`. The `Size_Of` function is mapped to a C++ function symbol `size_t size_of(Object_type const &)`. So the `Size_Of` functions for different types end up being overloads of `size_t size_of(..)`.

* At the construction time of the component, we add code like this:
  ```
  assert_valid_object_size<Request>();
  assert_valid_object_size<Primitive>();
  ```
  where the `Request` and `Primitive` are derived from `Spark::Object`.

* The `assert_valid_object_size` template function would look roughly like this:
  ```
  template <typename T>
  static inline void assert_valid_object_size()
  {
    if (Spark::size_of(*(T *)nullptr) > T::bytes())
      throw Object_size_mismatch();
  }
  ```

 The `Spark::Object<BYTES>` template must be extended with the `bytes` method with is a constexpr that returns the `BYTES` value. Of course, there could be also a diagnostic message.

  Granted, passing a nullptr as `in` argument into the `size_of` function looks a bit funny but I justify it with the following presumtions:
  1. The `Size_Of` functions are never supposed to actually access the argument. Only compile-time-generated meta data is returned. The object should remain untouched.
  2. The functions are called in this way only at start time of the component. So even if the Ada compiler generated unreasonable code that accesses the object, the code would produce a null-pointer access right at the start time of the component. Should that happen, we ought to reconsider this approach. ;-)

**Comment by m-stein** [comment]
Thanks, I'll give it a try.

**Comment by nfeske** [comment]
BTW, I tried out the idea by adding it to the test-spark example in https://github.com/genodelabs/genode/commit/a96d77504e4f2cffdac1f2c8a5085f8dd8ef252d

**Comment by m-stein** [comment]
My branch now contains the SPARK-part (no glue code) of the request pool. It does compile but is not tested so far:
https://github.com/m-stein/genode/commits/cbe

Next I'll integrate the suggestion by @nfeske and write the glue code for the request pool.

**Comment by cnuke** [comment]
Thanks, it looks good so far. I'm curious though, why you changed the signature of the `submit_request` method, it looks like the `Tag` is provided from the outside - do you intend to do that within the glue code?

**Comment by m-stein** [comment]
My preliminary idea is to hand over a tag from the outside that may come from a bit allocator in C++ or something alike. We could also do this allocation inside the pool module or simply cast the array index as before. I don't know but I wanted to get rid of the relation between the array index and the tag in the first place.

**Comment by nfeske** [comment]
I'd support the decoupling of tag's value from array indices.

In particular, it may be useful to just use the packet-stream offset of a request as tag. But then, the tag's values will span a large range, which makes it unsuitable for the use of array indices.

By treating it as an opaque value with the test for equality as the only operation, we are on the safe side.

**Comment by cnuke** [comment]
Okay, by changing the interface, it becomes possible to manage the tag from the outside, which makes it also opaque for the pool-module. That's fine and a proper justification for the interface change since now all modules are treated the same in this regard. (It is not necessary for removing the relation between array index and tag though, which is an implementation detail as I expressed in #5 if we treat a module like a black-box.)

**Comment by m-stein** [comment]
@nfeske The request-splitter module was implemented and later merged into the request pool. Therefore I would consider this issue as fixed. May I ask you to close it?
